Noisy test ...

Probably related to some randomness at some places...
```
======================================================================
FAIL: test_cython (fabio.test.test_densification.TestDensification)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/python/cp37-cp37m/lib/python3.7/site-packages/fabio/test/test_densification.py", line 73, in test_cython
    self.assertTrue(numpy.all(python == cython), "python == cython #" + str(i))
AssertionError: False is not true : python == cython #2
----------------------------------------------------------------------
```

--- a/fabio/ext/dense.pyx
+++ b/fabio/ext/dense.pyx
@@ -31,7 +31,7 @@
 """Densification of sparse frame format
 """
 __author__ = "Jérôme Kieffer"
-__date__ = "10/05/2021"  
+__date__ = "02/06/2022"  
 __contact__ = "Jerome.kieffer@esrf.fr"
 __license__ = "MIT"
 
@@ -44,7 +44,7 @@ from libc.stdint cimport int8_t, uint8_t, \
                          int64_t, uint64_t
 from libc.math cimport isfinite, log, sqrt, cos, M_PI                         
 from libc.stdlib cimport RAND_MAX
-cimport cython             
+cimport cython
                
 ctypedef fused any_t:
     double
@@ -221,15 +221,16 @@ def distribution_normal_mtc(mu, sigma, seed=None):
     return numpy.asarray(ary).reshape(shape)        
 
 
-def densify(float[:,::1] mask,
-            float[::1] radius,
+def densify(cython.floating[:,::1] mask,
+            cython.floating[::1] radius,
             uint32_t[::1] index,
             any_t[::1] intensity,
             any_t dummy,
             dtype,
             float[::1] background,
             float[::1] background_std=None,
-            normalization=None):
+            normalization=None,
+            seed = None):
     """
     Densify a sparse representation to generate a normal frame 
     
@@ -241,7 +242,8 @@ def densify(float[:,::1] mask,
     :param dummy: numerical value for masked-out pixels in dense image
     :param dtype: dtype of intensity.
     :param background_std: 1D array with the background std at given distance from the center --> activates the noisy mode.
-    :param 
+    :param normalization: normalization array: renormalize all data with this factor (pixel-wise)
+    :param seed: seed for the random number-generator, used only when regenerating noisy background
     :return: dense frame as 2D array
     """
     cdef:
@@ -267,15 +269,16 @@ def densify(float[:,::1] mask,
         noisy = False
     else:
         noisy=True
-        try:
-            value = time.time_ns()
-        except Exception:
-            value = int(time.time()*1e9)
-        mt = MT(value)
+        if seed is None:
+            try:
+                seed = time.time_ns()
+            except Exception:
+                seed = int(time.time()*1e9)
+        mt = MT(seed)
                 
     with nogil:
         start = radius[0]
-        idelta = (size - 1)/(radius[size-1] - start)  
+        idelta = <double>(size - 1)/(radius[size-1] - start)  
         
         #Linear interpolation
         for i in range(height):

--- a/fabio/sparseimage.py
+++ b/fabio/sparseimage.py
@@ -37,7 +37,7 @@ __authors__ = ["Jérôme Kieffer"]
 __contact__ = "jerome.kieffer@esrf.fr"
 __license__ = "MIT"
 __copyright__ = "2020 ESRF"
-__date__ = "10/05/2021"
+__date__ = "02/06/2022"
 
 import logging
 logger = logging.getLogger(__name__)
@@ -66,7 +66,8 @@ def densify(mask,
             dummy,
             background,
             background_std=None,
-            normalization=None):
+            normalization=None,
+            seed=None):
     """Generate a dense image of its sparse representation
     
     :param mask: 2D array with NaNs for mask and pixel radius for the valid pixels
@@ -82,6 +83,8 @@ def densify(mask,
     """
     dense = numpy.interp(mask, radius, background)
     if background_std is not None:
+        if seed is not None:
+            numpy.random.seed(seed)
         std = numpy.interp(mask, radius, background_std)
         numpy.maximum(0.0, numpy.random.normal(dense, std), out=dense)
     if normalization is not None:
@@ -91,7 +94,8 @@ def densify(mask,
     flat[index] = intensity
     dtype = intensity.dtype
     if numpy.issubdtype(dtype, numpy.integer):
-        dense = numpy.round(dense)
+        #dense = numpy.round(dense) # Foolded by banker's rounding !!!!
+        dense += 0.5
     dense = numpy.ascontiguousarray(dense, dtype=dtype)
     dense[numpy.logical_not(numpy.isfinite(mask))] = dummy
     return dense

--- a/fabio/test/test_densification.py
+++ b/fabio/test/test_densification.py
@@ -32,7 +32,7 @@ __authors__ = ["Jérôme Kieffer"]
 __contact__ = "Jerome.Kieffer@esrf.fr"
 __license__ = "MIT"
 __copyright__ = "2020 ESRF"
-__date__ = "28/04/2021"
+__date__ = "02/06/2022"
 
 import unittest
 import numpy
@@ -66,13 +66,15 @@ class TestDensification(unittest.TestCase):
         self.assertAlmostEqual(N.std(), sigma, 1)
 
     def test_cython(self):
+        seed = 0 
         shape = 256, 256
-        nframes = 8
+        nframes = 8000#
         vsize = 181  # This is cheated to avoid interpolation issues with rounding 128*sqrt(2)
         y, x = numpy.ogrid[-shape[0] // 2:-shape[0] // 2 + shape[0],
                           -shape[1] // 2:-shape[1] // 2 + shape[1]]
-        r2d = numpy.sqrt(x * x + y * y).astype(numpy.float32)
-        radius = numpy.linspace(0, r2d.max(), vsize).astype(numpy.float32)
+        # To make this test "robust", those two radial position arrays needs to be in float64 ... in production float32 is more common 
+        r2d = numpy.sqrt(x * x + y * y).astype(numpy.float64)
+        radius = numpy.linspace(0, r2d.max(), vsize).astype(numpy.float64)
         npeak = numpy.random.randint(90, 110, size=nframes)
         scale = numpy.random.randint(90, 110, size=nframes)
         osc = numpy.random.randint(40, 100, size=nframes)
@@ -92,6 +94,7 @@ class TestDensification(unittest.TestCase):
             f.ravel()[index[indptr[i]:indptr[i + 1]]] = intensity[indptr[i]:indptr[i + 1]]
             python = densify(r2d, radius, index[indptr[i]:indptr[i + 1]], intensity[indptr[i]:indptr[i + 1]], 0, background[i])
             cython = cython_densify.densify(r2d, radius, index[indptr[i]:indptr[i + 1]], intensity[indptr[i]:indptr[i + 1]], 0, intensity.dtype, background[i], None)
+            
             self.assertTrue(numpy.all(python == cython), "python == cython #" + str(i))
             # Rounding errors:
             delta = (python.astype(int) - f)
@@ -101,8 +104,8 @@ class TestDensification(unittest.TestCase):
             self.assertLess(len(bad[0]), numpy.prod(shape) / 500, "python differs from reference on less then 0.2% of the pixel #" + str(i))
 
             # Now consider the noise ...
-            python = densify(r2d, radius, index[indptr[i]:indptr[i + 1]], intensity[indptr[i]:indptr[i + 1]], 0, background[i], noise[i])
-            cython = cython_densify.densify(r2d, radius, index[indptr[i]:indptr[i + 1]], intensity[indptr[i]:indptr[i + 1]], 0, intensity.dtype, background[i], noise[i])
+            python = densify(r2d, radius, index[indptr[i]:indptr[i + 1]], intensity[indptr[i]:indptr[i + 1]], 0, background[i], noise[i], seed=seed)
+            cython = cython_densify.densify(r2d, radius, index[indptr[i]:indptr[i + 1]], intensity[indptr[i]:indptr[i + 1]], 0, intensity.dtype, background[i], noise[i], seed=seed)
             self.assertTrue(abs(python.astype(int) - cython).max() <= 2 * max(1, noise[i].max()), "python is close to cython #" + str(i))
 
 

--- a/fabio/test/test_densification.py
+++ b/fabio/test/test_densification.py
@@ -68,7 +68,7 @@ class TestDensification(unittest.TestCase):
     def test_cython(self):
         seed = 0 
         shape = 256, 256
-        nframes = 8000#
+        nframes = 8
         vsize = 181  # This is cheated to avoid interpolation issues with rounding 128*sqrt(2)
         y, x = numpy.ogrid[-shape[0] // 2:-shape[0] // 2 + shape[0],
                           -shape[1] // 2:-shape[1] // 2 + shape[1]]

